### PR TITLE
Added random sleep for given interval range. Fixes #9930

### LINKED ISSUE #9930: Allow Random Values for --sleep-interval
The `--sleep-interval` option is great. I think that this is the main reason why we can run downloads from heavily monitored sites.

Is there a way to modify this option so that it can accept random values? For example, instead of setting it to `--sleep-interval 300` to set it to sleep for 300 seconds or 5 minutes between each download, make it like so:

`--sleep-interval 120-300`

so that the sleep interval would be set between 120 and 300 seconds.

The rationale is to simulate more realistic browsing behavior.

### PR DESCRIPTION
**Pull request by singh-pratyush96**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] New extractor
- [x] New feature

### Added implementation that enables `--sleep-interval` to take interval range input

The input can be in the form `<lower limit>to<upper limit>` and `<upper limit>to<lower limit>` (where limits are in `seconds`), while keeping the original behavior (constant sleep) intact.

```
youtube-dl https://www.youtube.com/playlist?list=PLsPUh22kYmNDnezZ15MkicBGVDm2ypthG --sleep-interval 0to33e-1
```

As discusses at #9980, it is clear that the format `--sleep-interval a-b` won't work because of interval values like `2e-1`. Therefore using something like `to` would be a better option.

Right now, I have added `ValueError` to handle the case of faulty argument, which would in turn raise `UnavailableVideoError`. Though the name of the error is irrelevant to the actual error, I still used it because the `IOError` (raised from `time.sleep()`) for negative sleep interval in current version (`2016.07.24`) also raises it.

It would be better to add `SleepIntervalError` in `utils.py` that would provide a better verbose for user upon where things went wrong, instead of just showing `ERROR: unable to download video`.

### CODE REVIEW

File: youtube_dl/YoutubeDL.py [hunk 1]
```diff
@@ -250,6 +250,10 @@ class YoutubeDL(object):
     call_home:         Boolean, true iff we are allowed to contact the
                        youtube-dl servers for debugging.
     sleep_interval:    Number of seconds to sleep before each download.
+                        Accepts range for random sleep in formats:
+                        - <lower limit>to<upper limit>
+                        - <upper limit>to<lower limit>
+                            where limits are in seconds
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No, AtoB is definitely not how it should look like. It should be pair of options instead.

File: youtube_dl/YoutubeDL.py [hunk 2]
```diff
@@ -1684,7 +1688,7 @@ def compatible_formats(formats):
             except (compat_urllib_error.URLError, compat_http_client.HTTPException, socket.error) as err:
                 self.report_error('unable to download video data: %s' % error_to_compat_str(err))
                 return
-            except (OSError, IOError) as err:
+            except (OSError, IOError, ValueError) as err:
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No.

File: youtube_dl/downloader/common.py [hunk 3]
```diff
@@ -344,8 +345,16 @@ def download(self, filename, info_dict):
...
         if sleep_interval:
-            self.to_screen('[download] Sleeping %s seconds...' % sleep_interval)
-            time.sleep(sleep_interval)
+            sleep_interval_lower = sleep_interval.lower()
+
+            if 'to' in sleep_interval_lower:
+                lower_sleep_limit, upper_sleep_limit = map(float, sleep_interval_lower.split('to'))
+                sleep_time = random.uniform(lower_sleep_limit, upper_sleep_limit)
+            else:
+                sleep_time = float(sleep_interval)
```
Comment (dstftw) [inline youtube_dl/downloader/common.py]: All parsing and error reporting should take place in `__init__.py`.

File: README.md [hunk 4]
```diff
@@ -329,8 +329,11 @@ which means you can modify it, redistribute it or use it however you like.
     --bidi-workaround                Work around terminals that lack
                                      bidirectional text support. Requires bidiv
                                      or fribidi executable in PATH
-    --sleep-interval SECONDS         Number of seconds to sleep before each
-                                     download.
+    --sleep-interval SECONDS/RANGE   Number of seconds to sleep before each
+                                     download. Also accepts random sleep in 
+                                     format <lower limit>to<upper limit> and
+                                     <upper limit>to<lower limit>, where limits
+                                     are in seconds.
```
Comment (dstftw) [inline README.md]: This is autogenerated.

**Comment by singh-pratyush96** [comment]
Instead of having two options, will having another argument like `--sleep-offset` useful?

This offset would be used to select a random interval in `[a, a+b]`, where `a` is actual sleep interval and `b` is offset. This argument will be considered only if `--sleep-interval` is passed.

**Comment by dstftw** [comment]
By two options I mean `--sleep-interval` to be used as lower bound and to be mapped to `--min-sleep-interval`. The seconds option is `--max-sleep-interval` denotes the upper bound.

**Comment by singh-pratyush96** [comment]
HI!

I have implemented `--max-sleep-interval`. Please take a look.

File: youtube_dl/downloader/common.py [hunk 5]
```diff
@@ -344,8 +346,22 @@ def download(self, filename, info_dict):
...
+                        float(min_sleep_interval),
+                        float(max_sleep_interval)
+                    )
+                else:
+                    sleep_time = float(min_sleep_interval)
+
+                self.to_screen('[download] Sleeping %s seconds...' % sleep_time)
+                time.sleep(sleep_time)
+            except (ValueError, IOError) as err:
+                raise InvalidSleepTimeError()
```
Comment (dstftw) [inline youtube_dl/downloader/common.py]: I've already clarified there **must be no runtime checks here.** Sleep intervals are static information thus **must be validated on start**.

**Comment by dstftw** [comment]
Fix issues mentioned earlier and squash commits when done.

**Comment by singh-pratyush96** [comment]
Hi again!

I hope that this does well, except that I couldn't squash the commits into one from my side (failed as you can see). I guess that these commits can be squashed while merging.

**Comment by dstftw** [comment]
>Sleep intervals are static information thus **must be validated on start**.

https://www.google.ru/search?q=how+to+squash+commits

**Comment by singh-pratyush96** [comment]
> Sleep intervals are static information thus *must be validated on start*.

Doesn't the type of option as `float` validates the intervals on start? If no, what other checks are needed to be conducted?

**Comment by singh-pratyush96** [comment]
The solution provided for squashing was `rebase` which didn't removed the previous commits, but just added one more commit which reflected the changes from all other commits. Is this what was intended?

**Comment by dstftw** [comment]
You must ensure values are denoting a valid range 0<=a<=b.
What intended is a single commit without all these non relevant mess.

**Comment by singh-pratyush96** [comment]
Thanks for discussion. Will send another PR.

**Pull request closed without merging**